.github: Add pull request template (#7577)
NixOS/nix: .github/PULL_REQUEST_TEMPLATE.md

> ##### Motivation for the changes
> 
> <!-- An issue reference or explanation why the change is desirable. -->
> <!-- This speeds up step 1 of the process. -->
> 
> 
> 
> 
> 
> 
> ##### Checklist for maintainers

Reviewer: I think we should be more precise about what we want to have here.

```suggestion
# Motivation
<!-- Briefly explain what the change is about and why it is desirable. -->

# Context
<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Checklist for maintainers
```

While it's probably undesirable to have large headings in the description, it also doesn't really hurt to make the sections more distinct. After all, it's just a presentational aspect.

Reviewer 2: It may even be worthy to dedicate a whole (potentially blankable) section explicitly to design. It nudges people to possibly at least quickly revise their drawing board when they started to hack out  of an itch or momentum and feel really enthusiastic about their adrenaline rush.

```
# Design

<!-- Explain the implementation route and design that you've chosen and why. -->
```

Reviewer: That was my first idea, too, but that requires additional work for trivial PRs, which I think we'd want to have many of and which should require a trivial amount of effort. If the PR is complex anyway, authors may as well be inclined to make a proper write-up.

<!-- Contributors: please leave this as is -->
<!-- Maintainers: tick if completed or explain if not relevant -->

>  - [ ] is the idea good? has it been discussed by the Nix team?

Reviewer: ```suggestion
 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
```

To more closely match the [team handbook](https://github.com/NixOS/nix/tree/eaa20f25747eb025c7c4e69fa83f0e455a65393f/maintainers#project-board-protocol).

I thought about writing "approved", but that sounds a lot like bureaucracy.

Reviewer 2: `agreed on idea` &mdash;  `agreed on idea & design` I think the word "design" adds value around here.

 - [ ] unit tests
 - [ ] functional tests (`tests/**.sh`)

>  - [ ] documentation in the manual

Reviewer: ```suggestion
 - [ ] user-facing documentation is appropriate
```

We may want to be more precise about what "appropriate" means, maybe there is a better word...

Reviewer 2: It's a thing of balance, neither too verbose, nor too terse. It's hard to capture where to strike that balance without really having a clearly defined "user/reader persona" for which one is supposed to optimize.

It's a problem, indeed. "balanced" suffers from the same imprecision, but better captures that it's a judgemental trade off.

Author: I don't understand why this is an improvement.
It seems a bit wordy and we already have "tick if completed or explain if not relevant".

Reviewer: With the additional information in the other notes its indeed good as it is.

>  - [ ] documentation in the code (if necessary; ideally code is already clear)

Reviewer: ```suggestion
- [ ] code and comments are self-explanatory
```

I was trying to make it more tangible what "documentation in the code" is supposed to mean, but it's really hard to come up with something clear-cut.

I want code to be "readable" in a sense that if you read it, it answers all the questions about itself without having to dig around. That doesn't necessarily mean there need to be lots of comments. The symbol names should allow one to easily infer what they mean, comments should explain why the code is written the way it is, or explain what it does if there is no clean way of expressing it in the code itself. There is plenty of material out there that explains really well why these (and others) are qualities of readable code, and eventually we should collect these resources for reference.

# Motivation
<!-- Briefly explain what the change is about and why it is desirable. -->

# Context
<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

> <!-- Maintainers: tick if completed or explain if not relevant -->

Author: ```suggestion

Maintainers: tick if completed or explain if not relevant
```
